### PR TITLE
Implemented feature to import CSV to reset Inventory(#31a0f22)

### LINKED ISSUE #98: Implement feature to import CSV to reset Inventory
Once uploaded the inventory count will be set to the values in CSV.

Here is the format for CSV

facilityId | externalFacilityId | productId | idType | idValue | availableQty | locationSeqId | comments

### PR DESCRIPTION
**Pull request by disha1202**
### Related Issues

Closes #98

### Short Description and Why It's Useful

Implemented feature to import CSV, display product details and option to change facility and facility locations.

TODO: Need to add code to upload to oms.

### Screenshots of Visual Changes before/after (If There Are Any)

Screenshot from 2022-12-05 18-34-54
Screenshot from 2022-12-05 18-35-03
Screenshot from 2022-12-05 18-35-26
Screenshot from 2022-12-05 18-35-36

Reset Inventory CSV: [resetInventoryByIdentification.csv](https://github.com/hotwax/import/files/10674067/resetInventoryByIdentification.csv)

**IMPORTANT NOTICE** - Remember to add changelog entry

### Contribution and Currently Important Rules Acceptance

- [ ] I read and followed [contribution rules](https://github.com/hotwax/import#contribution-guideline)

### CODE REVIEW

File: src/store/modules/stock/actions.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        item.parentProductId = product.groupId;
+        item.internalName = product.internalName;
+        item.parentProductName = product.parentProductName;
+        item.imageUrl = product.mainImageUrl;
+        item.isNewProduct = "N";
+        item.isSelected = true;
+        return item;
+      }
+      unidentifiedProductItems.push(item);
+      return ;
```
Comment (k2maan) [inline src/store/modules/stock/actions.ts:34]: Improve code formatting.

File: src/views/Inventory.vue [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+</template>
+<script>
+import { IonPage, IonHeader, IonToolbar, IonTitle, IonContent, IonItem, IonLabel, IonList, IonListHeader, IonMenuButton, IonButton, IonSelect, IonSelectOption, IonIcon } from "@ionic/vue";
+import { defineComponent } from "vue";
+import { useRouter } from 'vue-router';
+import { useStore, mapGetters } from "vuex";
+import { showToast, parseCsv } from '@/utils';
+import { translate } from "@/i18n";
+import { arrowForwardOutline } from 'ionicons/icons';
+import { DateTime } from 'luxon'
```
Comment (k2maan) [inline src/views/Inventory.vue:67]: Remove unused import.

File: src/views/Inventory.vue [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    },
+    methods: {
+      getFile(event) {
+        this.file = event.target.files[0];
+        if(this.file){
+          showToast(translate("File uploaded successfully"));
+          this.parseFile();
+          this.store.dispatch('order/updateFileName', this.file.name);
+        }
+        else {
```
Comment (k2maan) [inline src/views/Inventory.vue:111]: I think we can improve code formatting here also

File: src/views/Inventory.vue [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+import { defineComponent } from "vue";
+import { useRouter } from 'vue-router';
+import { useStore, mapGetters } from "vuex";
+import { showToast, parseCsv } from '@/utils';
+import { translate } from "@/i18n";
+import { arrowForwardOutline } from 'ionicons/icons';
+import { DateTime } from 'luxon'
+
+export default defineComponent({
+    name: "purchase orders",
```
Comment (k2maan) [inline src/views/Inventory.vue:70]: Update component name

File: src/views/InventoryDetail.vue [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+        const resp = await OrderService.getFacilities(payload);
+        if(resp.status === 200 && !hasError(resp)){
+          this.facilities = resp.data.docs;
+        }
+      } catch(err) {
+        console.error(err)
+      }
+    },
+    async UpdateProduct(ev: Event, id: any, isVirtual: boolean, item: any, type: string) {
+      const popover = await popoverController
```
Comment (k2maan) [inline src/views/InventoryDetail.vue:318]: Use camelcase naming

File: src/views/OrderDetail.vue [hunk 6]
```diff
@@ -338,14 +338,14 @@ export default defineComponent({
         console.error(err)
       }
     },
-    async UpdateProduct(ev: Event, id: any, isVirtual: boolean, item: any) {
+    async UpdateProduct(ev: Event, id: any, isVirtual: boolean, item: any, type: string) {
```
Comment (k2maan) [inline src/views/OrderDetail.vue:341]: We can use camel casing here also

File: src/views/InventoryDetail.vue [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+import { useRouter } from 'vue-router';
+import { showToast } from '@/utils';
+import { translate } from "@/i18n";
+import { IonPage, IonHeader, IonToolbar, IonBackButton, IonContent, IonSearchbar, IonItem, IonThumbnail, IonLabel, IonChip, IonIcon, IonButton, IonCheckbox, IonSelect, IonSelectOption, IonButtons, popoverController, IonFab, IonFabButton, alertController, modalController } from '@ionic/vue'
+import { ellipsisVerticalOutline, locationOutline, checkboxOutline, cloudUploadOutline, arrowUndoOutline } from 'ionicons/icons'
+import { hasError } from "@/utils";
+import MissingSkuModal from "@/components/MissingSkuModal.vue"
+
+export default defineComponent({
+  name: 'PurchaseOrderDetail',
```
Comment (k2maan) [inline src/views/InventoryDetail.vue:148]: Update component name

File: src/store/modules/stock/getters.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import { GetterTree } from "vuex";
+import OrderState from "./StockState";
+import RootState from "../../RootState";
+
+const getters: GetterTree<OrderState, RootState> = {
+    getStock(state) {
+        return state.list;
```
Comment (k2maan) [inline src/store/modules/stock/getters.ts:7]: Fix indentation

File: src/views/Inventory.vue [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    },
+    setup() {
+    const router = useRouter();
+    const store = useStore();
+    return {
+      arrowForwardOutline,
+      router,
+      store
+    }
+  }
```
Comment (k2maan) [inline src/views/Inventory.vue:145]: Improve identation

File: src/views/InventoryDetail.vue [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+          <ion-label>{{ searchedProduct.quantity }} {{ $t("Items") }}</ion-label>
+        </ion-chip>
+
+        <ion-chip outline class="tablet">
+          <ion-label>{{ searchedProduct.externalFacilityId }}</ion-label>
+        </ion-chip>
+
+        <ion-chip outline class="tablet">
+          <ion-icon :icon="locationOutline" />
+          <ion-select interface="popover" :value="searchedProduct.locationSeqId" @ionChange="setFacilityLocation($event)">
```
Comment (k2maan) [inline src/views/InventoryDetail.vue:58]: Why are we passing only a single parameter in `setFacilityLocation` here?

File: src/views/Inventory.vue [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        </ion-list>
+
+        <ion-button color="medium" @click="mapFields" expand="block">
+          {{ $t("Review") }}
+          <ion-icon slot="end" :icon="arrowForwardOutline" />
+        </ion-button>
+      </main>
+    </ion-content>
+  </ion-page>
+</template>
```
Comment (shashwatbangar) [inline src/views/Inventory.vue:58]:
```suggestion
</template>

```

**Comment by shashwatbangar** [comment]
@disha1202 , verified workflow. **Looks good to me**. 

Just when **no inventory is received in API response** (with 200 status), **it shows error in console** which I think is not right. 

I suggest to display a toast saying something like "Inventory not found".

Screenshot 2023-01-04 at 11 20 03 AM

**Comment by disha1202** [comment]
> @disha1202 , verified workflow. **Looks good to me**.
> 
> Just when **no inventory is received in API response** (with 200 status), **it shows error in console** which I think is not right.
> 
> I suggest to **display a toast** saying something like "Inventory not found".
> 
> Screenshot 2023-01-04 at 11 20 03 AM

This will be handled differently and is not in scope for this PR.

**Review by shashwatbangar — APPROVED** [review]

File: src/views/InventoryDetail.vue [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+import Image from '@/components/Image.vue';
+import ProductPopover from '@/components/ProductPopover.vue'
+import { defineComponent } from 'vue';
+import { mapGetters, useStore } from "vuex";
+import { useRouter } from 'vue-router';
+import { showToast } from '@/utils';
+import { translate } from "@/i18n";
+import { IonPage, IonHeader, IonToolbar, IonBackButton, IonContent, IonSearchbar, IonItem, IonThumbnail, IonLabel, IonChip, IonIcon, IonButton, IonCheckbox, IonSelect, IonSelectOption, IonButtons, popoverController, IonFab, IonFabButton, alertController, modalController } from '@ionic/vue'
+import { ellipsisVerticalOutline, locationOutline, checkboxOutline, cloudUploadOutline, arrowUndoOutline } from 'ionicons/icons'
+import { hasError } from "@/utils";
```
Comment (ymaheshwari1) [inline src/views/InventoryDetail.vue:144]: Multiple imports from same file.

File: src/views/InventoryDetail.vue [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+        message: this.$t("Make sure all the data you have entered is correct."),
+        buttons: [
+            {
+              text: this.$t("Cancel"),
+              role: 'cancel',
+            },
+            {
+              text: this.$t("Upload"),
+              handler: () => {
+                const response = UploadService.uploadJsonFile(UploadService.prepareUploadJsonPayload({
```
Comment (ymaheshwari1) [inline src/views/InventoryDetail.vue:271]: response is not used, so I think we can remove this variable.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/views/InventoryDetail.vue (hunk 12 above)
Comment (disha1202) [reply src/views/InventoryDetail.vue]: Removed multiple imports from same file.

File: src/views/InventoryDetail.vue (hunk 13 above)
Comment (disha1202) [reply src/views/InventoryDetail.vue]: Removed unused variable.

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

File: src/locales/en.json [hunk 14]
```diff
@@ -1,6 +1,7 @@
 {
   "App": "App",
   "Apply": "Apply",
+  "Any edits made to this CSV will be lost.": "Any edits made to this CSV will be lost.",
```
[the remark this answers is not in the document]
Comment (disha1202) [reply src/locales/en.json]: Updated.

File: src/views/Inventory.vue [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+        productSku: "",
+        quantity: "",
+        facility: "",
+        locationSeqId: "",
+      },
+      productsList: [],
+    }
+  },
+  methods: {
+    getFile(event) {
```
[the remark this answers is not in the document]
Comment (disha1202) [reply src/views/Inventory.vue]: Created a mixin for redundant code.

File: src/store/modules/stock/StockState.ts [hunk 16]
```diff
@@ -0,0 +1,8 @@
+export default interface StockState { 
+  list: {
+    items: any,
+    original: any,
+    unidentifiedProductItems: any,
+  },
```
Comment (disha1202) [inline src/store/modules/stock/StockState.ts:6]:
```suggestion
  items: {
    parsed: any,
    original: any,
    unidentifiedItems: any,
  },  
```

File: src/views/InventoryDetail.vue [hunk 17]
```diff
@@ -0,0 +1,442 @@
...
+
+          <div />
+          
+          <ion-checkbox :checked="isParentProductChecked(id)" @click="isParentProductUpdated = true" @ionChange="selectParentProduct(id, $event)" />
+
+          <ion-button fill="clear" color="medium" @click="updateProduct($event, id, true, getParentInformation(id, stock.items), 'stock')">
+            <ion-icon slot="icon-only" :icon="ellipsisVerticalOutline" />
+          </ion-button>
+        </div>
+        <div v-for="(item, index) in getGroupItems(id, stock.items)" :key="index">
```
[the remark this answers is not in the document]
Comment (disha1202) [reply src/views/InventoryDetail.vue:89]: The method name should be `getItemsByParentProduct` or `getItemsByVirtualProduct`.
@adityasharma7 Sir WDYT?

File: src/views/InventoryDetail.vue [hunk 18]
```diff
@@ -0,0 +1,442 @@
...
+          </ion-button>
+        </ion-buttons>
+      </ion-toolbar>
+    </ion-header>
+
+    <ion-content >
+      <div class="header">
+        <div class="search">
+          <ion-searchbar  :placeholder="$t('Search products')" v-model="queryString" v-on:keyup.enter="queryString = $event.target.value; searchProduct(queryString)"></ion-searchbar>
+          <ion-chip outline @click="listMissingSkus()">
```
Comment (disha1202) [inline src/views/InventoryDetail.vue:21]:
```suggestion
          <ion-chip outline @click="openMissingSkuModal()">
```

File: src/store/modules/stock/getters.ts [hunk 19]
```diff
@@ -0,0 +1,10 @@
+import { GetterTree } from "vuex";
+import OrderState from "./StockState";
+import RootState from "../../RootState";
+
+const getters: GetterTree<OrderState, RootState> = {
+  getStock(state) {
```
Comment (disha1202) [inline src/store/modules/stock/getters.ts:6]:
```suggestion
  getItemsStock(state) {
```

File: src/mixins/parseFileMixin.ts [hunk 20]
```diff
@@ -0,0 +1,20 @@
+import { translate } from "@/i18n";
+import store from "@/store";
+import { showToast, parseCsv } from "@/utils";
+
+export default {
+  methods: {
+    async parseCsv(file: any) {
+      if (file) {
+          showToast(translate("File uploaded successfully"));
```
Comment (k2maan) [inline src/mixins/parseFileMixin.ts:9]: Improve indentation

File: src/store/modules/stock/actions.ts [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+      if(Object.keys(product).length > 0){
+        item.parentProductId = product?.parent?.id;
+        item.pseudoId = product.pseudoId;
+        item.parentProductName = product?.parent?.productName;
+        item.imageUrl = product.images?.mainImageUrl;
+        item.isSelected = true;
+        return item;
+      }
+      unidentifiedItems.push(item);
+      return ;
```
Comment (k2maan) [inline src/store/modules/stock/actions.ts:33]: Remove unwanted spacing.

File: src/views/PurchaseOrder.vue [hunk 22]
```diff
@@ -127,44 +129,40 @@ export default defineComponent({
...
-          showToast(translate("Select all the fields to continue"));
+          return;
         } 
+
+        if (!this.areAllFieldsSelected()) {
+          showToast(translate("Select all the fields to continue"));
+          return;
+        }
+
+        this.PurchaseOrderItems = this.content.map(item => {
```
Comment (k2maan) [inline src/views/PurchaseOrder.vue:152]: We can use lower camel casing to name variables.

File: src/components/BulkInventoryAdjustmentModal.vue [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+  },
+  data() {
+    return {
+      quantityBuffer: "",
+      facilityId: "",
+    }
+  },
+  computed: {
+    ...mapGetters({
+      stock: 'stock/getItemsStock',
```
Comment (adityasharma7) [inline src/components/BulkInventoryAdjustmentModal.vue:83]: Why not?
```suggestion
      stockItems: 'stock/getItems',  
```

File: src/components/MissingFacilitiesModal.vue [hunk 24]
```diff
@@ -79,23 +79,40 @@ export default defineComponent({
...
-    save(){
-      Object.keys(this.facilityMapping).map((facilityId: any) => {
-        Object.values(this.purchaseOrders.parsed).flat().map((item: any) => {
-          if(item.externalFacilityId === facilityId){
-            item.externalFacilityId = "";
-            item.facilityId = this.facilityMapping[facilityId];
-          }
+    async save(){
+      if(this.type === 'order'){
+        Object.keys(this.facilityMapping).map((facilityId: any) => {
```
[the remark this answers is not in the document]
Comment (adityasharma7) [reply src/components/MissingFacilitiesModal.vue:92]: Or a separate method called here

File: src/components/MissingSkuModal.vue [hunk 25]
```diff
@@ -152,22 +157,29 @@ export default defineComponent({
...
+      unidentifiedItem.updatedSku = this.updatedSku;
+      unidentifiedItem.parentProductId = item.parent.id;
+      unidentifiedItem.pseudoId = item.pseudoId;
+      unidentifiedItem.parentProductName = item.parent.productName;
+      unidentifiedItem.imageUrl = item.images.mainImageUrl;
+      unidentifiedItem.isSelected = true;
+  
+      this.hasSkuUpdated = true;
+      if (this.type === 'order'){
+        unidentifiedItem.isProductNew = 'N';
```
Comment (adityasharma7) [inline src/components/MissingSkuModal.vue:179]:
```suggestion
        unidentifiedItem.isNewProduct = 'N';
```

File: src/services/UserService.ts [hunk 26]
```diff
@@ -44,6 +44,14 @@ const setUserTimeZone = async (payload: any): Promise <any>  => {
   });
 }
 
+const getFacilityLocations = async (payload: any): Promise<any> => {
+  return api({
+    url: "/performFind",
+    method: "POST",
```
Comment (adityasharma7) [inline src/services/UserService.ts:50]: We could use GET method

File: src/services/UserService.ts [hunk 27]
```diff
@@ -77,13 +85,14 @@ const getFieldMappings = async (payload: any): Promise <any> => {
...
-    createFieldMapping,
-    deleteFieldMapping,
-    login,
-    getAvailableTimeZones,
-    getFieldMappings,
-    getProfile,
-    setUserTimeZone,
-    checkPermission,
-    updateFieldMapping
+  createFieldMapping,
```
Comment (adityasharma7) [inline src/services/UserService.ts:88]: Unwanted changes

File: src/store/modules/stock/actions.ts [hunk 28]
```diff
@@ -0,0 +1,63 @@
+import { ActionTree } from 'vuex'
+import store from '@/store'
+import RootState from '@/store/RootState'
+import StockState from './StockState'
+import * as types from './mutation-types'
+
+const actions: ActionTree<StockState, RootState> = {
+  async updatedStockList ({commit, rootGetters}, items) {
```
Comment (adityasharma7) [inline src/store/modules/stock/actions.ts:8]:
```suggestion
  async updateStock ({commit, rootGetters}, items) {
```

File: src/store/modules/stock/actions.ts [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+    const payload = {
+      viewSize,
+      viewIndex,
+      productIds
+    }
+    await store.dispatch('user/fetchFacilityLocations', facilityIds);
+    await store.dispatch("product/fetchProducts", payload);
+    const unidentifiedItems = [] as any;
+    const parsed = items.map((item: any) => {
+      const product = rootGetters['product/getProduct'](item.shopifyProductSKU)
```
Comment (adityasharma7) [inline src/store/modules/stock/actions.ts:22]: We should not use getter here, instead get products list with product/fetchProducts action and then use that

File: src/store/modules/stock/actions.ts [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+    }).filter((item: any) => item);
+
+    const original = JSON.parse(JSON.stringify(parsed));
+
+    commit(types.STOCK_ITEMS_UPDATED, { parsed, original, unidentifiedItems });
+  },
+  updateStockItems({ commit }, stockItems){
+    commit(types.STOCK_ITEMS_UPDATED, stockItems);
+  },
+  clearStockList({ commit }){
```
Comment (adityasharma7) [inline src/store/modules/stock/actions.ts:43]:
```suggestion
  clearStock({ commit }){
```

File: src/store/modules/user/UserState.ts [hunk 31]
```diff
@@ -5,6 +5,7 @@ export default interface UserState {
     instanceUrl: string;
     fieldMappings: object;
     preferredDateTimeFormat: string;
+    facilityLocationsByFacilityId: any;
```
Comment (adityasharma7) [inline src/store/modules/user/UserState.ts:8]: I think this should be part of util state instead

File: src/views/Inventory.vue [hunk 32]
```diff
@@ -0,0 +1,163 @@
...
+    IonMenuButton,
+    IonSelect,
+    IonSelectOption,
+    IonIcon,
+    IonListHeader,
+    IonList
+  },
+  computed: {
+    ...mapGetters({
+      dateTimeFormat : 'user/getDateTimeFormat',
```
Comment (adityasharma7) [inline src/views/Inventory.vue:89]: I think this is not used

File: src/views/Inventory.vue [hunk 33]
```diff
@@ -0,0 +1,163 @@
...
+          shopifyProductSKU: item[this.fieldMapping.productSku],
+          quantity: item[this.fieldMapping.quantity],
+          facilityId: '',
+          externalFacilityId: item[this.fieldMapping.facility],
+          locationSeqId: item[this.fieldMapping.locationSeqId]
+        }
+      })
+      this.store.dispatch('stock/updatedStockList', this.productsList);
+      this.router.push({
+        name:'InventoryDetail'
```
Comment (disha1202) [inline src/views/Inventory.vue:134]:
```suggestion
        name:'InventoryReview'
```

File: src/views/Inventory.vue [hunk 34]
```diff
@@ -0,0 +1,163 @@
...
+      this.file = event.target.files[0];
+      this.content = await this.parseCsv(this.file);
+    },
+    mapFields() {
+      const areAllFieldsSelected = Object.values(this.fieldMapping).every(field => field !== "");
+      if (this.content.length <= 0) {
+        showToast(translate("Please upload a valid reset inventory csv to continue"));
+        return;
+      }
+
```
Comment (adityasharma7) [inline src/views/Inventory.vue:117]:
```suggestion
      if (this.content.length <= 0) {
        showToast(translate("Please upload a valid reset inventory csv to continue"));
        return;
      }
      
      const areAllFieldsSelected = Object.values(this.fieldMapping).every(field => field !== "");
```

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

**Pull request merged**